aten/src/ATen/ThreadLocalState.h

```diff
@@ -38,6 +39,9 @@ class TORCH_API ThreadLocalState {
   // RecordFunction TLS
   RecordFunctionTLS rf_tls_;
 
+  // TLS for functorch
+  std::shared_ptr<functorch::FuncTorchTLSBase> functorch_tls_;
```

Comment:
This has probably been answered before somewhere, but why is it that we need to store the state in a shared_ptr, vs. e.g. `AutogradState` can just be stored directly? We need the pointer because of the virtual methods, but is `shared_ptr` preferred over a raw pointer?

Reply:
Yes, we need a pointer because of the virtual methods. I didn't use a raw pointer because C++ style guides generally [prefer smart pointers (shared_ptr, unique_ptr, etc) to raw pointers](https://www.oreilly.com/library/view/effective-modern-c/9781491908419/ch04.html). The reason being that there is automatic memory management so that smart pointers eliminate an entire class of errors where you malloc() something and then forget to free it.

Now, the automatic memory management here doesn't matter too too much because we're storing the smart pointer in a static thread_local variable that is going to last the duration of the entire program. But it saves us some lines of code from needing to use malloc to allocate memory from the struct.

Reply:
I'm not sure to follow your argument here?
In particular because this wasn't needed for the autograd state and the corresponding tls: https://github.com/pytorch/pytorch/blob/master/c10/core/AutogradState.cpp ?

Reply:
The reason why we need a pointer is because we want to extend FunctorchTLSBase inside functorch. Inside functorch, we will create a subclass of FunctorchTLSBase called FunctorchTLSImpl that actually contains metadata, like the DynamicLayerStack. I don't want to put the metadata into FunctorchTLSBase because the metadata hasn't been defined yet.

If we instead stored a `FunctorchTLSBase functorch_tls_` in ThreadLocalState, then we would not be able to assign FunctorchTLSImpl to `functorch_tls_`. This is because `sizeof(FunctorchTLSImpl) > sizeof(FunctorchTLSBase)` (due to there being additional metadata). Performing the assignment of a FunctorchTLSImpl to `FunctorchTLSBase functorch_tls_` does a truncation which is not good.

Now, the reason why we need a smart pointer rather than a raw pointer is that smart pointers are better at memory management and I don't want to manually call malloc.

Reply:
(If that explanation makes sense I can drop it as a comment into the code)

Reply:
It does sound good. It would indeed be nice to have a comment in the code to justify why only this one is a pointer.

Reply:
I'd love to make this a unique_ptr, but ThreadLocalState is copyable and gets copied somewhere in the codebase. Making it a unique_ptr makes it so that compiler rightfully complains about ThreadLocalState being copied